src/compute/src/server.rs

```diff
@@ -168,6 +171,7 @@ pub async fn compute_node_serve(
     let worker_id = meta_client.worker_id();
     info!("Assigned worker node id {}", worker_id);
 
+    // TODO(shutdown): remove this as there's no need to gracefully shutdown the sub-tasks.
```

Comment:
I don't fully understand it. The PR's description said:

> ... Instead, directly shutdown the runtime when the main task finds all done.

So, how does the main task find all things done, without joining these subtasks (i.e. a bunch of `Managers`)?

Reply:
Since (by examining the code) we know that these sub-tasks don't do anything meaningful when they're asked to shutdown, so we don't have to join them.

If there's really something that's meaningful, it's true that we still need to join the sub-task. But I suppose it can be more ad-hoc, instead of always collecting all sub-tasks into a vec.

Reply:
I am neutral for this point. Personally, I'd like to join every sub-tasks in a hierarchy style, even most of them are doing _effectively_ nothing - just some useless work such as stopping some background tasks.

Reply:
Unfortunately we have a lot more detached tasks that are spawned but not recorded as sub-tasks everywhere in our codebase. This is natural, because some of them are simple enough and exit behavior is quite manageable (like polling the receiver side of a channel and exit when the sender is closed). As a result, tracking every of them will lead to a hell of maintenance as described by @kwannoel in Slack:

> Agree, we have lots of shutdown channels to graceful shutdown different managers in meta.
It’s quite high overhead to maintain all these.

Given that it's hard to make sure we've covered all spawned tasks here, it'll still be impossible to thoroughly shutdown the runtime (instead of `shutdown_background` in this PR) so we still have to exit the process manually. So I'm afraid there's no much benefits for doing this. 😕 I feel like consciously synchronizing with the tasks that really need some sort of graceful shutdown is sufficient.